Add compiler directive to disable the default exception propagation (#5094)
cython/cython: docs/src/userguide/source_files_and_compilation.rst

    Copy the original source code line by line into C code comments in the generated
    code file to help with understanding the output.
    This is also required for coverage analysis.

``cpp_locals`` (True / False)
    Make C++ variables behave more like Python variables by allowing them to be
    "unbound" instead of always default-constructing them at the start of a
    function.  See :ref:`cpp_locals directive` for more detail.

> ``legacy_implicit_noexcept`` (True / False)

Author: I am not native english speaker so I can be cumbersome. I would like to somehow discourage users using this directive at all. Can we mention somehow that this directive is just for migration. I was thinking about mentioning that this directive is temporary for migration purposes and will be removed in the future?

Reviewer: I think:

> This directive is intended to help migrate legacy code written before Cython 3. We recommend leaving it off for new code.

I don't think there's a need to threaten to remove it yet.

Author: thanks!

Reviewer 2: I only saw this part of the discussion now. I think it's ok to tell users that it will eventually be removed. That's what "deprecation" means, really. We may find a wording that keeps them from perpetually asking "are we there yet", though.

Author: Commited @scoder suggestion with message of future removal. Marking as resolved.

>     When enabled, ``cdef`` functions will not propagate raised exception by default. Hence,

Reviewer: ```suggestion
    When enabled, ``cdef`` functions will not propagate raised exceptions by default. Hence,
```

    Copy the original source code line by line into C code comments in the generated
    code file to help with understanding the output.
    This is also required for coverage analysis.

``cpp_locals`` (True / False)
    Make C++ variables behave more like Python variables by allowing them to be
    "unbound" instead of always default-constructing them at the start of a
    function.  See :ref:`cpp_locals directive` for more detail.

``legacy_implicit_noexcept`` (True / False)
    When enabled, ``cdef`` functions will not propagate raised exceptions by default. Hence,

>     the function will behave same way as if declared with `noexcept` keyword. See

Reviewer 2: ```suggestion
    the function will behave in the same way as if declared with `noexcept` keyword. See
```

    :ref:`error_return_values` for details. Setting this directive to ``True`` will

>     cause Cython to have same semantics as Cython 0.2X. This directive is intended to help
>     migrate legacy code written before Cython 3. We recommend leaving it off for new code.

Reviewer 3: Since it already raises a deprecation warning, I think this statement could be strengthened:
```suggestion
    cause Cython to have same semantics as Cython 0.2X. This directive is deprecated and
    exists solely to help migrate legacy code written before Cython 3.
```

Author: Not sure about it. This text was added after proposal by @da-woods

Reviewer: I like "exists solely to help..." - that seems like a mild improvement.

It feels a bit odd to deprecate something before you even add it so I'd possibly rather not use the word "deprecated".

Reviewer: Although we use the word "deprecated" elsewhere so maybe its fine.

Author: I think this can be marked as resolved after @scoder suggestion

>     cause Cython to have same semantics as Cython 0.2X. This directive is intended to help
>     migrate legacy code written before Cython 3. We recommend leaving it off for new code.

Reviewer 2: ```suggestion
    cause Cython 3.0 to have the same semantics as Cython 0.x. This directive was solely added
    to help migrate legacy code written before Cython 3. It will be removed in a future release.
```
